ci/eval: add darwin support

winterqt opened the pull request:
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

winterqt committed:
ci/eval: properly kill mem stats job on exit

Before this change, the eval derivations would never actually complete
on any platform other than sandboxed Linux because the mem stats job
would never be terminated. For reasons I don't know, the Linux sandbox
somehow results in Bash terminating the job on its own. On every other
platform (and when the Linux sandbox is disabled), we have to do it
correctly and kill it manually.

winterqt committed:
ci/eval: add darwin support

winterqt commented:
Eval failure is expected until #406303 is merged.

No idea why the code owner action can't request y'all's reviews, maybe because it's a draft? GH's API doesn't document such a limitation, though...

Happy to split the last commit into a different PR -- in fact, it may make sense to off the bat. Let me know.

lilyball commented on ci/eval/default.nix:
```diff
@@ -152,8 +160,16 @@ let
+            ${if stdenv.isDarwin then ''
+              freeBlocks=$(vm_stat | grep free | awk '{ print $3 }' | sed 's/\.//')
+              inactiveBlocks=$(vm_stat | grep inactive | awk '{ print $3 }' | sed 's/\.//')
+              speculativeBlocks=$(vm_stat | grep speculative | awk '{ print $3 }' | sed 's/\.//')
```
Minor suggestion, we can do a single call to `vm_stat` and also do the grepping in awk:

```bash
output=$(vm_stat | sed 's/\.//g')
freeBlocks=$(awk '/free/ { print $3 }' <<<$output)
inactiveBlocks=$(awk '/inactive/ { print $3 }' <<<$output)
speculativeBlocks=$(awk '/speculative/ { print $3 }' <<<$output)
```

lilyball commented on ci/eval/default.nix:
```diff
@@ -152,8 +160,16 @@ let
+              freeBlocks=$(vm_stat | grep free | awk '{ print $3 }' | sed 's/\.//')
+              inactiveBlocks=$(vm_stat | grep inactive | awk '{ print $3 }' | sed 's/\.//')
+              speculativeBlocks=$(vm_stat | grep speculative | awk '{ print $3 }' | sed 's/\.//')
+              availMemory=$((($freeBlocks+$inactiveBlocks+$speculativeBlocks)*$(pagesize)))
```
Minor nit, but math expressions can refer to variables directly, they don't need to be interpolated:

```bash
availMemory=$(( (freeBlocks+inactiveBlocks+speculativeBlocks) * $(pagesize) ))
```

lilyball commented on ci/eval/default.nix:
```diff
@@ -152,8 +160,16 @@ let
+              inactiveBlocks=$(vm_stat | grep inactive | awk '{ print $3 }' | sed 's/\.//')
+              speculativeBlocks=$(vm_stat | grep speculative | awk '{ print $3 }' | sed 's/\.//')
+              availMemory=$((($freeBlocks+$inactiveBlocks+$speculativeBlocks)*$(pagesize)))
+              freeSwap=$(($(sysctl vm.swapusage | awk '{ print $10; }' | sed 's/M//' | sed -E 's/\..+//') * 1024 * 1024))
```
Minor nit, the seds can be removed by just asking awk to print the value as an integer

```bash
freeSwap=$(( $(sysctl vm.swapusage | awk '{ printf "%d", $10 }') * 1024 * 1024 ))
```

lilyball reviewed (commented):
Just some minor comments on the bash scripting

winterqt commented:
Read the error in the later run, it tells you to set `sandbox` configuration option to `relaxed` — that’s what you have to do to get this to work.

wolfgangwalther reviewed (approved):
According to the latest run above, this works on both x86_64-darwin and aarch64-darwin.

I'm thinking that we might want to have some way to confirm this in CI every now and then - eval should keep working on all platforms, so that all contributors can run it locally, too. I have a few ideas, but we don't need to do this in here.

winterqt commented:
Rebased. Not sure if we should wait till `eval.full` can be fixed on Darwin via bumping the pinned Nixpkgs, or if we should just do it.

winterqt commented:
Dropped the `x86_64-linux` eval system change in favor of waiting for #406825, but everything else should be good to go now that I've updated the pin.

winterqt commented on ci/eval/default.nix:
```diff
@@ -74,6 +82,28 @@ let
+      abortCmd =
+        if stdenv.isDarwin then
+          ''
+            # For some reason, `kill` (or `kill -9`) won't
+            # actually kill an entire process tree on macOS
+            # as it does on Linux. So instead, we'll just
+            # use xargs' native termination feature, which
+            # will at least guarantee that the iteration
+            # will be stopped.
+            #
+            # (Yes, technically we can just keep the `kill`
+            # command so that xargs dies and then the pending
+            # chunks will finish, but I'd rather just do this
+            # if it's going to have the same effect anyways.)
+            exit 255
```
Maybe there is a way to get this to work if there's a way to get the parent's process group ID? Not sure.

lilyball replied:
As discussed on Discord, I believe this should actually just be `kill 0` for all platforms. AFAICT `kill $PPID` only works on linux because of the sandbox (that causes the command to fail with an error which `set -e` causes the builder to exit, and presumably that causes the pid namespace to get torn down), xargs does not proactively kill its spawned processes when it dies.

If you do use `kill 0`, that signals the entire process group, which is going to include the shell itself since it's a non-interactive shell (which means job control defaults to off). If you want just the xargs process tree to shut down you can enable job control first with `set -m`, that way the xargs pipeline will be its own process group and the `kill 0` will only affect it, at which point bash's `set -e` behavior will take over. So doing this really just depends on whether there's any benefits to log output (or whether Nix cares at all about whether the builder exits with a code or a signal).

winterqt replied:
> AFAICT `kill $PPID` only works on linux because of the sandbox

This would match the issue with the memory stats job that I observed and also bisected down to the Linux sandbox. Fun.

Thanks, I’ll give that a shot. I don’t think we want to `set -m` at least without somehow explicitly detecting whether the `xargs` pipeline was killed or not (which I’m not sure is possible?).

lilyball replied:
`set -m` just enables job control monitor mode, which makes every pipeline get its own process group (and bash will print a line when background jobs exit). It shouldn't actually hurt anything, it just makes bash behave a bit more like interactive mode. and you can presumably turn it off again after the xargs finishes if you like.

I'm not sure what you mean about explicitly detecting whether the `xargs` pipeline was killed? If job control is enabled, the `kill 0` will just kill `xargs` (and `seq` if it hasn't already finished) and its child processes, and bash will treat that as an error and exit (due to `set -e`). If you want to intercept that, you can wrap the xargs pipeline in an `if` or add a `||` after and test the status code (and a code over 128 indicates a signal, where the signal number is the exit status minus 128, e.g. SIGTERM is 143). But normal errexit behavior should be fine here? It's what you get with `kill $PPID` already.

lilyball commented on ci/eval/default.nix:
```diff
@@ -163,6 +209,8 @@ let
           done
         ) &
 
+        trap "kill %%" EXIT
```
This isn't going to work if anyone adds another background job later on, since the jobspec here is evaluated at EXIT time instead of right now.

```suggestion
        trap "kill $(jobs -p %%)" EXIT
```

Also I just want to point out that stdenv installs a default exit handler that this overwrites. That may not matter at all, if nothing installs a `failureHook` or `exitHook`, but if you want to preserve that you could switch this over to adding both a `failureHook` and `exitHook` (or just a `failureHook` and adding an explicit kill after the `xargs` finishes).

wolfgangwalther reviewed (commented):
> Dropped the x86_64-linux eval system change in favor of waiting for https://github.com/NixOS/nixpkgs/pull/406825, but everything else should be good to go now that I've updated the pin.

Do we still need to wait for #406825 or has the required change maybe already been merged (since I have been splitting off smaller parts repeatedly)?

Is there anything else, especially from the latest comments, that you'd want to address / change in here?

We just had #418153, too, so making eval work on darwin would be valuable to others, too.
